statistics/histogram.go

```diff
@@ -858,6 +867,156 @@ func (idx *Index) getRowCount(sc *stmtctx.StatementContext, indexRanges []*range
 	return totalCount, nil
 }
 
+type countByRangeFunc = func(*stmtctx.StatementContext, int64, []*ranger.Range) (float64, error)
+
+// newHistogramBySelectivity fulfills the content of new histogram by the given selectivity result.
+// TODO: Datum is not efficient, try to avoid using it here.
+//  Also, there're redundant calculation with Selectivity(). We need to reduce it too.
+func newHistogramBySelectivity(sc *stmtctx.StatementContext, histID int64, oldHist, newHist *Histogram, ranges []*ranger.Range, cntByRangeFunc countByRangeFunc) error {
+	splitRanges := oldHist.SplitRange(ranges, false)
+	cntPerVal := int64(oldHist.AvgCountPerValue(int64(oldHist.totalRowCount())))
+	var totCnt int64 = 0
+	for boundIdx, ranIdx, highRangeIdx := 0, 0, 0; boundIdx < oldHist.Bounds.NumRows() && ranIdx < len(splitRanges); boundIdx, ranIdx = boundIdx+2, highRangeIdx {
+		for highRangeIdx < len(splitRanges) && chunk.Compare(oldHist.Bounds.GetRow(boundIdx+1), 0, &splitRanges[highRangeIdx].HighVal[0]) >= 0 {
+			highRangeIdx++
+		}
+		if boundIdx+2 >= oldHist.Bounds.NumRows() && highRangeIdx < len(splitRanges) && splitRanges[highRangeIdx].HighVal[0].Kind() == types.KindMaxValue {
+			highRangeIdx++
+		}
+		if ranIdx == highRangeIdx {
+			continue
+		}
+		cnt, err := cntByRangeFunc(sc, histID, splitRanges[ranIdx:highRangeIdx])
+		// This should not happen.
+		if err != nil {
+			return err
+		}
+		if cnt == 0 {
+			continue
+		}
+		if int64(cnt) > oldHist.bucketCount(boundIdx/2) {
+			cnt = float64(oldHist.bucketCount(boundIdx / 2))
+		}
+		newHist.Bounds.AppendRow(oldHist.Bounds.GetRow(boundIdx))
+		newHist.Bounds.AppendRow(oldHist.Bounds.GetRow(boundIdx + 1))
+		totCnt += int64(cnt)
+		bkt := Bucket{Count: totCnt}
+		if chunk.Compare(oldHist.Bounds.GetRow(boundIdx+1), 0, &splitRanges[highRangeIdx-1].HighVal[0]) == 0 && !splitRanges[highRangeIdx-1].HighExclude {
+			bkt.Repeat = cntPerVal
+		}
+		newHist.Buckets = append(newHist.Buckets, bkt)
+		switch newHist.Tp.EvalType() {
+		case types.ETString, types.ETDecimal, types.ETDatetime, types.ETTimestamp:
+			newHist.scalars = append(newHist.scalars, oldHist.scalars[boundIdx/2])
+		}
+	}
+	return nil
+}
+
+func (idx *Index) newIndexBySelectivity(sc *stmtctx.StatementContext, statsNode *StatsNode) (*Index, error) {
+	var (
+		ranLowEncode, ranHighEncode []byte
+		err                         error
+	)
+	newIndexHist := &Index{Info: idx.Info}
```

Comment:
There are still other things need to be maintained, like `statsVersion`, `CM Sketch`.

Reply:
`stats version` is useless during plan optimization?

Reply:
It is used to decide the estimation methods for the index.

Reply:
PTAL @winoros

Reply:
And safe for index at line 938